src/Index/Index.php

```diff
@@ -216,10 +287,53 @@ public function unserialize($serialized)
     public function serialize()
     {
         return serialize([
-            'definitions' => $this->definitions,
+            'namespaceDefinitions' => $this->namespaceDefinitions,
             'references' => $this->references,
             'complete' => $this->complete,
             'staticComplete' => $this->staticComplete
         ]);
     }
+
+    /**
+     * Registers a definition to the global definitions index if it is global
+     *
+     * @param string $fqn The fully qualified name of the symbol
+     * @param Definition $definition The Definition object
+     * @return void
+     */
+    private function setGlobalDefinition(string $fqn, Definition $definition)
+    {
+        if ($definition->isGlobal) {
+            $this->globalDefinitions[$fqn] = $definition;
+        }
+    }
+
+    /**
+     * @param string $fqn
+     * @return string The namespace extracted from the given FQN
+     */
+    private function extractNamespace(string $fqn): string
+    {
+        foreach (['::', '->'] as $operator) {
+            if (false !== ($pos = strpos($fqn, $operator))) {
+                return substr($fqn, 0, $pos);
+            }
+        }
+
+        return $fqn;
+    }
+
+    /**
+     * Returns the Definitions that are in the given namespace
+     *
+     * @param string $namespace
+     * @return Definition[]
+     */
+    private function doGetDefinitionsForNamespace(string $namespace): array
```

Comment:
Don't fully see the need for this since it's only `$this->namespaceDefinitions[$namespace] ?? []` but if you want to keep it please rename it to something like `getDefinitionArrayForNamespace` - the current name has no semantic difference to the non-`do` version